### PR TITLE
992/public primary unit group

### PR DESCRIPTION
**Pull request by seanmalbert**
## Issue

- Addresses #992
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This adds the primary unit group summary to the listing detail page.

Screen Shot 2022-03-07 at 10 47 18 PM

A few things to note while reviewing this:
1. for the sake of time and because there are other unit group summaries to still be built I did not abstract helper functions where we will likely want to in both the front end and backend
2. I did not include the info icon next to AMI, because that requires an update to the StandardTable element as well as the info icon and I think that should be done separately

This also adds household maximum income.

Screen Shot 2022-03-08 at 2 23 09 PM

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Prototype/POC (not to merge)
- [ ] This change is a refactor/addresses technical debt
- [ ] This change requires a documentation update
- [ ] This change requires a SQL Script

## How Can This Be Tested/Reviewed?

Pull this down and run `yarn db:reseed:detroit`. On the public side, find the listing titled `MARTIN LUTHER KING II`. Go to the detail, and you should see the table.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have updated the changelog to include a description of my changes
- [ ] I have run `yarn generate:client` if I made backend changes
- [ ] I have exported any new pieces in ui-components

### CODE REVIEW

File: backend/core/src/units/types/household-max-income-summary.ts [hunk 1]
```diff
@@ -1,9 +1,10 @@
 import { ApiProperty } from "@nestjs/swagger"
+import { HMIColumns } from "./household-max-income-columns"
 
 export class HouseholdMaxIncomeSummary {
   @ApiProperty()
-  columns: { [key: string]: unknown }
+  columns: HMIColumns
```
Comment (emilyjablonski) [inline backend/core/src/units/types/household-max-income-summary.ts:6]: 🥳

File: sites/public/src/ListingView.tsx [hunk 2]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
     availability: t("t.availability"),
+    ami: t("listings.unit.ami"),
   }
 
   let groupedUnits: Record<string, React.ReactNode>[] = null
-  if (listing.unitsSummary !== undefined && listing.unitsSummary.length > 0) {
-    groupedUnits = getSummariesTableFromUnitsSummary(listing.unitsSummary)
-  } else if (listing.unitsSummarized !== undefined) {
-    groupedUnits = getSummariesTableFromUnitSummary(listing.unitsSummarized.byUnitTypeAndRent)
+  let hmiHeaders: TableHeaders
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:79]: I'd love to consider pulling this into sites/public/lib/helpers since this file is already long and hard to parse

File: sites/public/src/ListingView.tsx [hunk 3]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+      let rentAsPercentIncomeRange = null
+      if (group.rentRange && group.rentRange.min === group.rentRange.max) {
+        rentRange = group.rentRange.min
+      } else if (group.rentRange) {
+        rentRange = `${group.rentRange.min} - ${group.rentRange.max}`
+      }
+
+      if (rentRange) {
+        rentRange = (
+          <span>
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:94]: Can we just do this on line 89

File: sites/public/src/ListingView.tsx [hunk 4]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+          </span>
+        )
+      }
+
+      if (
+        group.rentAsPercentIncomeRange &&
+        group.rentAsPercentIncomeRange.min === group.rentAsPercentIncomeRange.max
+      ) {
+        rentAsPercentIncomeRange = group.rentAsPercentIncomeRange.min
+      } else if (rentAsPercentIncomeRange) {
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:105]: `group.rentAsPercentIncomeRange`

File: sites/public/src/ListingView.tsx [hunk 5]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+        rentAsPercentIncomeRange = `${group.rentAsPercentIncomeRange.min} - ${group.rentAsPercentIncomeRange.max}`
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:106]: Can we have percentages on both numbers here? `30% - 40%` instead of `30 - 40%`

File: sites/public/src/ListingView.tsx [hunk 6]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+      }
+
+      if (rentAsPercentIncomeRange) {
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:109]: Same as above, can we do this on 106 and remove this block

File: sites/public/src/ListingView.tsx [hunk 7]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+      }
+
+      let availability = null
+
+      if (group.unitVacancies > 0) {
+        availability = (
+          <div>
+            <strong>{group.unitVacancies} </strong>
+            {group.unitVacancies === 1 ? t("lisitngs.vacantUnit") : t("listings.vacantUnits")}
+            {" &"}
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:124]: We will want to translate `&`

File: sites/public/src/ListingView.tsx [hunk 8]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+        )
+      }
+
+      let availability = null
+
+      if (group.unitVacancies > 0) {
+        availability = (
+          <div>
+            <strong>{group.unitVacancies} </strong>
+            {group.unitVacancies === 1 ? t("lisitngs.vacantUnit") : t("listings.vacantUnits")}
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:123]: `listings.vacantUnit`

File: sites/public/src/ListingView.tsx [hunk 9]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+      )
+
+      let ami = null
+
+      if (
+        group.amiPercentageRange &&
+        group.amiPercentageRange.min === group.amiPercentageRange.max
+      ) {
+        ami = `${group.amiPercentageRange.min}%`
+      } else {
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:145]: Do we want to check if `group.amiPercentageRange` isn't null here again as we did above?

File: sites/public/src/ListingView.tsx [hunk 10]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+        ami = `${group.amiPercentageRange.min} - ${group.amiPercentageRange.max}%`
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:146]: Can we have percentages on both numbers here? 30% - 40% instead of 30 - 40%

File: sites/public/src/ListingView.tsx [hunk 11]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+      }
+
+      return {
+        unitType: (
+          <>
+            {group.unitTypes
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:152]: We should sort by numBedrooms

File: sites/public/src/ListingView.tsx [hunk 12]
```diff
@@ -74,13 +72,145 @@ export const ListingView = (props: ListingProps) => {
...
+              .map<React.ReactNode>((type) => (
+                <strong key={type}>{t(`listings.unitTypes.${type}`)}</strong>
+              ))
+              .reduce((acc, curr) => [acc, ", ", curr])}
+          </>
+        ),
+        rent:
+          rentRange || rentAsPercentIncomeRange ? (
```
Comment (emilyjablonski) [inline sites/public/src/ListingView.tsx:160]: I find this a little hard to read, here is one suggestion:
let rentCol = ""
if (rentRange) rentCol = rentRange
if (rentAsPercentIncomeRange) rentCol = `${rentRange}${rentRange.length && ", "}${rentAsPercentIncomeRange}

File: ui-components/src/locales/general.json [hunk 13]
```diff
@@ -1382,6 +1382,10 @@
       "studio": "Studio",
       "SRO": "SRO"
     },
+    "vacantUnit": "vacant unit",
+    "vacantUnits": "vacant units",
+    "unitSummaryGroupMessage": "For each type of unit, you cannot make more than the income limit shown for the",
```
Comment (emilyjablonski) [inline ui-components/src/locales/general.json:1387]: Ideally we don't concat strings like this for translations where these concats won't make sense

File: backend/core/src/auth/filters/user-query-filter.ts [hunk 14]
```diff
@@ -52,3 +53,28 @@ export class UserQueryFilter extends BaseQueryFilter {
     }
   }
 }
+
+export function addIsPortalUserQuery(qb: WhereExpression, filterValue: string) {
```
[the remark this answers is not in the document]
Comment (seanmalbert) [reply backend/core/src/auth/filters/user-query-filter.ts:57]: 986 isn't up-to-date with dev, so updating that will remove a lot of the changes seen in this PR.

File: sites/public/src/ListingView.tsx (hunk 10 above)
Comment (seanmalbert) [reply sites/public/src/ListingView.tsx:146]: This follows the mocks, which I think I agree with, but not so strongly that I'd argue against a consensus to change it.

**Comment by seanmalbert** [comment]
> The tables aren't showing on the directory page, should they be here? 

The issue doesn't cover the directory and it's not needed for this deadline, since partner users will only be able to see the listings via the preview.

> Can we add one default unit group to all other listings so something shows? Or not show the table headers if there are no unit groups?

I'll update to not show if there are no unit groups.

> Should we add tests for the units-transformations fxns?
Yes, trying to get things out iteratively.

File: sites/public/src/ListingView.tsx (hunk 3 above)
Comment (seanmalbert) [reply sites/public/src/ListingView.tsx:94]: We'd also have to do it on line 87, if we did that.

File: sites/public/src/ListingView.tsx (hunk 11 above)
Comment (seanmalbert) [reply sites/public/src/ListingView.tsx:152]: unitTypes are strings here, so adding the sort in the backend within units-transformatins

File: sites/public/src/ListingView.tsx (hunk 12 above)
Comment (seanmalbert) [reply sites/public/src/ListingView.tsx:160]: Updating to:
```
let rent: React.ReactNode = null

if (rentRange && rentAsPercentIncomeRange) {
  rent = (
    <>
      {rentRange}, {rentAsPercentIncomeRange}
    </>
  )
} else if (rentRange) {
  rent = rentRange
} else if (rentAsPercentIncomeRange) {
  rent = rentAsPercentIncomeRange
}
```

File: ui-components/src/locales/general.json (hunk 13 above)
Comment (seanmalbert) [reply ui-components/src/locales/general.json:1388]: I agree. I'm going to leave it like this for now and think of a better way to handle links within translated text.

**Comment by seanmalbert** [comment]
@emilyjablonski ,
I addressed most of your concerns. We still need tests for unit transformations and ideally for the new helpers in public, but I'm going to handle those in a separate PR. Can you take another look please.
